src/compiler/binder.ts

```diff
@@ -2362,6 +2362,11 @@ namespace ts {
                     const symbolTable = hasModifier(thisContainer, ModifierFlags.Static) ? containingClass.symbol.exports : containingClass.symbol.members;
                     declareSymbol(symbolTable, containingClass.symbol, node, SymbolFlags.Property, SymbolFlags.None, /*isReplaceableByMethod*/ true);
                     break;
+                case SyntaxKind.SourceFile:
+                    // this.foo assignment in a source file
+                    // Bind this property in the global namespace
+                    declareSymbolAndAddToSymbolTable(node, SymbolFlags.FunctionScopedVariable, SymbolFlags.FunctionScopedVariableExcludes);

```

Comment:
The following test doesn't work perfectly:
```ts
// @out: output.js
// @allowJs: true
// @Filename: a.js
this.a = 10;
a;

// @Filename: /b.js
this.a;
a;
```

In `/b.js`, `this.a` will be `any`, while `a` is still `number`.
I don't see how this ends up in globals? `declareSourceFileMember` seems to put it in the source file locals, but does that really point to the globals instead?

Reply:
this seems half working, we either acknowledge that `this` at the top level is the global scope, and mirror that both for declarations and look ups, or we do not do either.

in other words, if `this.a = 10` is a declaration of a global `a`, then should `var a = 10;` and both should be visible on `this.a`.

For this bug fix, i would say just avoid the assert by handling the `SourceFile` case.

Reply:
Note that top level `this` in a commonjs file is the initial `exports` object.

`this` is only `global` in a traditional namespace-style file.

Reply:
## Binding

@andy-ms the single file case works the same way. `file.locals` declarations on a script (not module) source file will be available in the global namespace. As @Kovensky points out, we should put the declaration on `file.exports` when the source file is a module.

## Type checking

@mhegazy Regarding the type checking, I guess you want us to start treating `this` as `{ [s: string]: any, a: number }` instead of `any`? For the module case, we can treat it as the type of `exports`, which is actually the source file, I think. I guess we can look up members on top level this by looking them up in file.locals or file.exports, and just returning any if the member is not found.

## Short term fix
@mhegazy It’s easy enough to no-op source files — do you think a short-term fix is worth it here? I guess the customer that @sheetalkamat was working with would benefit. I’ll create a separate PR with just the no-op and keep this one open for the complete fix.